File: src/services/mailbox-rotation.ts
import { PrismaClient, MailboxStatus } from '@prisma/client';

const prisma = new PrismaClient();

export interface MailboxSelectionOptions {
  tenantId: string;
  excludeMailboxIds?: string[];
  requireMinimumSends?: number;
}

export interface SelectedMailbox {
  id: string;
  email: string;
  displayName?: string;
  weight: number;
  dailySendLimit: number;
  sentToday: number;
  availableSends: number;
}

export class MailboxRotationService {
  /**
   * Select a mailbox for sending using weighted selection
   */
  async selectMailbox(options: MailboxSelectionOptions): Promise<SelectedMailbox | null> {
    const { tenantId, excludeMailboxIds = [], requireMinimumSends = 1 } = options;

    // Get all active mailboxes for the tenant
    const mailboxes = await prisma.mailbox.findMany({
      where: {
        tenantId,
        status: MailboxStatus.ACTIVE,
        id: {
          notIn: excludeMailboxIds,
        },
        // Only include mailboxes that haven't reached their daily limit
        sentToday: {
          lt: prisma.mailbox.fields.dailySendLimit,
        },
      },
      select: {
        id: true,
        email: true,
        displayName: true,
        weight: true,
        dailySendLimit: true,
        sentToday: true,
      },
    });

    if (mailboxes.length === 0) {
      return null;
    }

    // Calculate available sends and filter by minimum requirement
    const availableMailboxes = mailboxes
      .map(mailbox => ({
        ...mailbox,
        availableSends: mailbox.dailySendLimit - mailbox.sentToday,
      }))
      .filter(mailbox => mailbox.availableSends >= requireMinimumSends);

    if (availableMailboxes.length === 0) {
      return null;
    }

    // Perform weighted selection
    const selectedMailbox = this.performWeightedSelection(availableMailboxes);
    return selectedMailbox;
  }

  /**
   * Perform weighted random selection
   */
  private performWeightedSelection(mailboxes: SelectedMailbox[]): SelectedMailbox {
    // Calculate total weight, adjusting for availability
    const totalWeight = mailboxes.reduce((sum, mailbox) => {
      // Boost weight based on available capacity
      const availabilityBoost = mailbox.availableSends / mailbox.dailySendLimit;
      return sum + (mailbox.weight * availabilityBoost);
    }, 0);

    // Generate random number
    const random = Math.random() * totalWeight;
    
    // Find the selected mailbox
    let currentWeight = 0;
    for (const mailbox of mailboxes) {
      const availabilityBoost = mailbox.availableSends / mailbox.dailySendLimit;
      currentWeight += (mailbox.weight * availabilityBoost);
      
      if (random <= currentWeight) {
        return mailbox;
      }
    }

    // Fallback to the first mailbox (shouldn't happen)
    return mailboxes[0];
  }

  /**
   * Reserve a send slot for a mailbox
   */
  async reserveSend(mailboxId: string): Promise<boolean> {
    try {
      const result = await prisma.mailbox.updateMany({
        where: {
          id: mailboxId,
          sentToday: {
            lt: prisma.mailbox.fields.dailySendLimit,
          },
        },
        data: {
          sentToday: {
            increment: 1,
          },
        },
      });

      return result.count > 0;
    } catch (error) {
      console.error('Failed to reserve send slot:', error);
      return false;
    }
  }

  /**
   * Release a reserved send slot (in case of send failure)
   */
  async releaseSend(mailboxId: string): Promise<boolean> {
    try {
      const result = await prisma.mailbox.updateMany({
        where: {
          id: mailboxId,
          sentToday: {
            gt: 0,
          },
        },
        data: {
          sentToday: {
            decrement: 1,
          },
        },
      });

      return result.count > 0;
    } catch (error) {
      console.error('Failed to release send slot:', error);
      return false;
    }
  }

  /**
   * Get mailbox statistics for a tenant
   */
  async getMailboxStats(tenantId: string) {
    const mailboxes = await prisma.mailbox.findMany({
      where: { tenantId },
      select: {
        id: true,
        email: true,
        displayName: true,
        status: true,
        weight: true,
        dailySendLimit: true,
        sentToday: true,
      },
    });

    const stats = {
      total: mailboxes.length,
      active: mailboxes.filter(m => m.status === MailboxStatus.ACTIVE).length,
      paused: mailboxes.filter(m => m.status === MailboxStatus.PAUSED).length,
      outOfOffice: mailboxes.filter(m => m.status === MailboxStatus.OUT_OF_OFFICE).length,
      bounced: mailboxes.filter(m => m.status === MailboxStatus.BOUNCED).length,
      disabled: mailboxes.filter(m => m.status === MailboxStatus.DISABLED).length,
      totalDailyLimit: mailboxes.reduce((sum, m) => sum + m.dailySendLimit, 0),
      totalSentToday: mailboxes.reduce((sum, m) => sum + m.sentToday, 0),
      availableCapacity: mailboxes
        .filter(m => m.status === MailboxStatus.ACTIVE)
        .reduce((sum, m) => sum + (m.dailySendLimit - m.sentToday), 0),
    };

    return {
      stats,
      mailboxes: mailboxes.map(m => ({
        ...m,
        availableSends: m.dailySendLimit - m.sentToday,
        utilizationRate: (m.sentToday / m.dailySendLimit * 100).toFixed(1),
      })),
    };
  }

  /**
   * Reset daily send counters for all mailboxes
   */
  async resetDailySendCounters(tenantId?: string): Promise<number> {
    const where = tenantId ? { tenantId } : {};
    
    const result = await prisma.mailbox.updateMany({
      where,
      data: {
        sentToday: 0,
      },
    });

    return result.count;
  }

  /**
   * Check circuit breaker conditions
   */
  async checkCircuitBreaker(tenantId: string, timeWindow = 24): Promise<{
    shouldTrigger: boolean;
    bounceRate: number;
    totalEmails: number;
    bounces: number;
  }> {
    const cutoffTime = new Date(Date.now() - (timeWindow * 60 * 60 * 1000));

    // Get total emails sent in the time window
    const totalEmails = await prisma.sentEmail.count({
      where: {
        mailbox: {
          tenantId,
        },
        sentAt: {
          gte: cutoffTime,
        },
      },
    });

    if (totalEmails === 0) {
      return {
        shouldTrigger: false,
        bounceRate: 0,
        totalEmails: 0,
        bounces: 0,
      };
    }

    // Get bounce signals in the same time window
    const bounces = await prisma.signal.count({
      where: {
        type: 'BOUNCE',
        mailbox: {
          tenantId,
        },
        detectedAt: {
          gte: cutoffTime,
        },
      },
    });

    const bounceRate = (bounces / totalEmails) * 100;
    const shouldTrigger = bounceRate > 2.0; // 2% threshold

    return {
      shouldTrigger,
      bounceRate: parseFloat(bounceRate.toFixed(2)),
      totalEmails,
      bounces,
    };
  }

  /**
   * Pause mailboxes due to circuit breaker
   */
  async triggerCircuitBreaker(tenantId: string, reason = 'High bounce rate detected'): Promise<number> {
    const result = await prisma.mailbox.updateMany({
      where: {
        tenantId,
        status: MailboxStatus.ACTIVE,
      },
      data: {
        status: MailboxStatus.PAUSED,
      },
    });

    // TODO: Create webhook event for circuit breaker trigger
    
    return result.count;
  }
}

export const mailboxRotationService = new MailboxRotationService();

File: src/services/mailbox_rotation.py
"""Mailbox rotation service for weighted selection."""

import random
from datetime import datetime, timedelta, timezone

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from src.models.models import Mailbox, MailboxStatus
from src.config import TENANT_MAILBOX_MAP, ALL_ALLOWED_MAILBOXES

# Daily capacity (sent_today) is reset at 00:05 UTC (com.scout.mailbox-reset /
# the arq reset cron). When a mailbox is at the hard 75/day cap we defer the
# send to just after this reset rather than crashing. See next_capacity_reset.
CAPACITY_RESET_HOUR = 0
CAPACITY_RESET_MINUTE = 5


def next_capacity_reset(now: datetime | None = None) -> datetime:
    """Return the next UTC datetime at which daily send capacity resets (00:05 UTC).

    The seq-service capacity counter (Mailbox.sent_today) is zeroed at 00:05 UTC.
    A worker that finds its sticky mailbox at the hard cap re-queues its step to
    just after this reset instead of raising/retrying. If now is already inside
    the 00:00 to 00:05 window the reset is the SAME day's 00:05; otherwise it is
    the NEXT day's 00:05. now defaults to the current UTC time.
    """
    if now is None:
        now = datetime.now(timezone.utc)
    reset_today = now.replace(
        hour=CAPACITY_RESET_HOUR, minute=CAPACITY_RESET_MINUTE,
        second=0, microsecond=0,
    )
    if now < reset_today:
        return reset_today
    return reset_today + timedelta(days=1)


def seconds_until_capacity_reset(now: datetime | None = None) -> int:
    """Whole seconds from now until the next 00:05 UTC capacity reset (>=1)."""
    if now is None:
        now = datetime.now(timezone.utc)
    delta = (next_capacity_reset(now) - now).total_seconds()
    return max(1, int(delta))


async def select_mailbox(
    db: AsyncSession,
    tenant_id: str,
    exclude_ids: list[str] | None = None,
    min_available: int = 1,
) -> Mailbox | None:
    """
    Select a mailbox using weighted random selection.

    Considers:
    - Daily send limit vs sent today
    - Mailbox weight
    - Active status only
    - HARDCODED tenant mailbox allocation (cannot be bypassed)

    Returns None when no active, allowed mailbox has at least `min_available`
    capacity remaining. None is the clean "at-capacity / unavailable" signal the
    caller maps to a defer/429 (it never raises for capacity).
    """
    exclude_ids = exclude_ids or []

    # Get active mailboxes with available capacity
    query = (
        select(Mailbox)
        .where(
            Mailbox.tenant_id == tenant_id,
            Mailbox.status == MailboxStatus.ACTIVE,
            Mailbox.id.notin_(exclude_ids),
        )
    )

    result = await db.execute(query)
    mailboxes = result.scalars().all()

    # HARDCODED ENFORCEMENT: Only allow mailboxes in the tenant's allocation
    allowed_emails = TENANT_MAILBOX_MAP.get(tenant_id, ALL_ALLOWED_MAILBOXES)
    mailboxes = [m for m in mailboxes if m.email in allowed_emails]

    # Filter by available sends
    available = [
        m for m in mailboxes
        if (m.daily_send_limit - m.sent_today) >= min_available
    ]

    if not available:
        return None

    # Weighted random selection
    weights = []
    for mailbox in available:
        # Boost weight based on remaining capacity
        capacity_ratio = (mailbox.daily_send_limit - mailbox.sent_today) / mailbox.daily_send_limit
        adjusted_weight = mailbox.weight * (1 + capacity_ratio)
        weights.append(adjusted_weight)

    selected = random.choices(available, weights=weights, k=1)[0]
    return selected


async def reserve_send(db: AsyncSession, mailbox_id: str) -> bool:
    """
    Atomically reserve a send slot.

    F4: a single conditional UPDATE (sent_today += 1 WHERE sent_today <
    daily_send_limit) instead of SELECT-then-increment, so concurrent workers
    can never both read the same sent_today and over-send past the daily limit.
    rowcount == 1 means we got a slot; 0 means the mailbox was at capacity (or
    not found).

    The bool return is the clean at-capacity signal: False == "at the hard cap,
    no slot". Callers DEFER (re-queue to the next 00:05 reset) on False; the API
    layer maps it to 429. This function never raises for capacity, and the
    atomic 75/day enforcement is unchanged.
    """
    result = await db.execute(
        update(Mailbox)
        .where(
            Mailbox.id == mailbox_id,
            Mailbox.sent_today < Mailbox.daily_send_limit,
        )
        .values(sent_today=Mailbox.sent_today + 1)
    )
    await db.commit()
    return result.rowcount == 1


async def release_send(db: AsyncSession, mailbox_id: str) -> None:
    """
    Release a previously-reserved send slot (sent_today -= 1, floored at 0).

    F5: reserve_send runs before the Gmail send so we respect the cap up front,
    but a failed send must NOT permanently consume capacity. Callers release the
    slot when the send fails. Floored at 0 so a double-release never goes negative.
    """
    await db.execute(
        update(Mailbox)
        .where(Mailbox.id == mailbox_id, Mailbox.sent_today > 0)
        .values(sent_today=Mailbox.sent_today - 1)
    )
    await db.commit()


async def reset_all_sent_today(db: AsyncSession, tenant_id: str) -> int:
    """Reset sent_today for all mailboxes in a tenant. Returns count updated."""
    result = await db.execute(
        update(Mailbox)
        .where(Mailbox.tenant_id == tenant_id)
        .values(sent_today=0)
    )
    await db.commit()
    return result.rowcount
